### PR TITLE
Add support for zstd dictionaries

### PR DESCRIPTION
**Pull request by miam-miam100**
zstd dictionaries allow you to train on similar data to achieve better compression ratios on small files. This PR provides no way for people to train their own data, just use pre-trained dictionaries for decoding and encoding.

### CODE REVIEW

File: src/codec/zstd/decoder.rs [hunk 1]
```diff
@@ -14,6 +14,13 @@ impl ZstdDecoder {
             decoder: Unshared::new(Decoder::new().unwrap()),
         }
     }
+
+    pub(crate) fn new_with_dict(dictionary: &[u8]) -> Self {
+        let mut decoder = Decoder::with_dictionary(dictionary).unwrap();
```
Comment (NobodyXu) [inline src/codec/zstd/decoder.rs:19]: loading dict could fail, so unwrapping here isn't a good idea.

File: src/macros.rs [hunk 2]
```diff
@@ -98,8 +122,24 @@ macro_rules! algos {
             }
         });
 
-        algos!(@algo xz ["xz"] XzDecoder XzEncoder<$inner> {
-            pub fn with_quality(inner: $inner, level: crate::Level) -> Self {
+        algos!(@algo-dec zstd ["zstd"] ZstdDecoder<$inner_dec> {
```
Comment (NobodyXu) [inline src/macros.rs:125]: Why can't you write 

```suggestion
impl<R> ZstdDecoder<R> {
```

then you won't need to change the macros.

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: src/codec/zstd/decoder.rs (hunk 1 above)
Comment (miam-miam100) [reply src/codec/zstd/decoder.rs:19]: No problem been changed, is it guaranteed that `new` and `new_with_params` can't panic?

File: src/macros.rs (hunk 2 above)
Comment (miam-miam100) [reply src/macros.rs:125]: Bcs then we would need to specify the trait R implements
```rs
impl<R: tokio::io::AsyncBufRead> ZstdDecoder<R> {
``` 
which is dependent on the async executor.

Comment (NobodyXu) [reply src/macros.rs:125]: Thanks, you are right in this

**Review by NobodyXu — APPROVED** [review]

**Comment by NobodyXu** [comment]
@robjtede Can you review this please?
Since this involves changes to macro_rules, I'd like to review from another maintainer.

**Comment by NobodyXu** [comment]
@robjtede Can you review this and get this into main before the next minor release?

**Comment by robjtede** [comment]
Reviewing this was very helpful in gaining a better understanding of how the macros work. So, I've made some simplifications that reduce noise in the diff. While giving the same API diff:

```
 Documenting async-compression v0.4.0 (/Users/rob/Development/forks/async-compression)
    Finished dev [unoptimized + debuginfo] target(s) in 0.48s
    Blocking waiting for file lock on package cache
 Documenting async-compression v0.4.0 (/Users/rob/Development/forks/async-compression)
    Finished dev [unoptimized + debuginfo] target(s) in 0.53s
Removed items from the public API
=================================
(none)

Changed items in the public API
===============================
(none)

Added items to the public API
=============================
+pub fn async_compression::futures::bufread::ZstdDecoder<R>::with_dict(inner: R, dictionary: &[u8]) -> std::io::error::Result<Self>
+pub fn async_compression::futures::bufread::ZstdEncoder<R>::with_dict(inner: R, level: async_compression::Level, dictionary: &[u8]) -> std::io::error::Result<Self>
+pub fn async_compression::futures::write::ZstdDecoder<W>::with_dict(inner: W, dictionary: &[u8]) -> std::io::error::Result<Self>
+pub fn async_compression::futures::write::ZstdEncoder<W>::with_dict(inner: W, level: async_compression::Level, dictionary: &[u8]) -> std::io::error::Result<Self>
+pub fn async_compression::tokio::bufread::ZstdDecoder<R>::with_dict(inner: R, dictionary: &[u8]) -> std::io::error::Result<Self>
+pub fn async_compression::tokio::bufread::ZstdEncoder<R>::with_dict(inner: R, level: async_compression::Level, dictionary: &[u8]) -> std::io::error::Result<Self>
+pub fn async_compression::tokio::write::ZstdDecoder<W>::with_dict(inner: W, dictionary: &[u8]) -> std::io::error::Result<Self>
+pub fn async_compression::tokio::write::ZstdEncoder<W>::with_dict(inner: W, level: async_compression::Level, dictionary: &[u8]) -> std::io::error::Result<Self>


```

**Comment by robjtede** [comment]
Note on simplification:

The actual problem to be solved was adding custom constructors to the decoders _as well_ as encoders. So instead of poluting the type parameters at callsites, I've just added a way in the macro to do so, with clear markers.

Future work: the runtime encoder/decoder macros know what their generics should be. It it probbaly possible to omit the $inner param from the algos! macro.

**Review by robjtede — APPROVED** [review]

**Comment by robjtede** [comment]
Added some testing for this feature.

@NobodyXu if you're happy with my changes we can merge this.

**Comment by robjtede** [comment]
Oh great... Windows-only test failures.

**Comment by miam-miam100** [comment]
Thanks @robjtede, sorry I couldn't help I've been competing in the GMTK game jam

File: tests/artifacts/lib.rs [hunk 3]
```diff
@@ -0,0 +1,352 @@
+//! Adaptors between compression crates and Rust's modern asynchronous IO types.
+//!
```
Comment (NobodyXu) [inline tests/artifacts/lib.rs:2]: What is this file for?

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: tests/artifacts/lib.rs (hunk 3 above)
Comment (robjtede) [reply tests/artifacts/lib.rs:2]: it _is_ the test case

Comment (NobodyXu) [reply tests/artifacts/lib.rs:2]: But this file `tests/artifacts/lib.rs` contains duplicate code from [`src/lib.rs`](https://github.com/Nullus157/async-compression/blob/main/src/lib.rs).

Did you accidentally copied `src/lib.rs` to `testrs/artifacts/lib.rs`?

Comment (NobodyXu) [reply tests/artifacts/lib.rs:2]: Oh I see, it's the file being compressed.
I thought it was some duplicate code in tests and it got me confused.

**Review by NobodyXu — APPROVED** [review]

**Comment by NobodyXu** [comment]
P.S. @robjtede I think using squash merge is better than merge commits, especially in this PR where I merge changes in main back to this PR.

With squash merge, the commit history would be much more straight-forward.

If you agree, I will disable merge commit in setting and only allow squash merging.

**Comment by NobodyXu** [comment]
I've disabled merge commit and rebase merge, allowing only squash merge.

**Pull request merged**